### PR TITLE
Refs #17405. Refactoring of CompareSampleLogs.

### LINKED ISSUE #17405: Refactor CompareSampleLogs
### Expected behavior
Experience shows, that sometimes numeric sample logs are represented as strings. However, users expect them to be compared as numbers (with certain tolerance).

The aim of the refactoring is to implement the behaviour expected by users, e.g. compare temperatures as numbers, even if they have type 'String'.

### Platforms affected
All

Please, assign me.

### PR DESCRIPTION
**Pull request by mganeva**
Refactoring of ComateSampleLogs algorithm to fix the mentioned issue and some minor code improvements. 

2 new unit tests.

**To test:**
Check that nothing has got broken. The simple test for new behavoiur:

``` python
ws1 = CreateSampleWorkspace()
ws2 = CreateSampleWorkspace()
AddSampleLog(ws1, 'chopper_speed', '13995', 'String')
AddSampleLog(ws2, 'chopper_speed', '14004', 'Number')
result = CompareSampleLogs([ws1, ws2], 'chopper_speed', 10.0)
print result
```

the result must be an empty string.

Fixes #17405. 

Does not need to be in the release notes.

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [x] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?
##### Functional Tests
- [x] Do changes function as described? Add comments below that describe the tests performed?
- [x] How do the changes handle unexpected situations, e.g. bad input?
- [x] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by DavidFair** [comment]
LGTM :shipit:

File: Framework/PythonInterface/plugins/algorithms/CompareSampleLogs.py [hunk 1]
```diff
@@ -39,8 +41,8 @@ def PyInit(self):
 
         self.declareProperty(StringArrayProperty(name="SampleLogs", direction=Direction.Input, validator=validator),
                              doc="Comma separated list of sample logs to compare.")
-        self.declareProperty("Tolerance", 1e-3, validator=FloatBoundedValidator(lower=1e-7, upper=1.0),
-                             doc="Tolerance for comparison of double values.")
+        self.declareProperty("Tolerance", 1e-3, validator=FloatBoundedValidator(lower=1e-7, upper=100.0),
```
Comment (AndreiSavici) [inline Framework/PythonInterface/plugins/algorithms/CompareSampleLogs.py]: Why 100?

Comment (mganeva) [inline Framework/PythonInterface/plugins/algorithms/CompareSampleLogs.py]: TOFTOF chopper speeds can differ around 10. 100 is just to give some freedom. May be better would be to remove the upper limit, but I don't know any property which could have reasonable tolerance more than 100.

Comment (AndreiSavici) [inline Framework/PythonInterface/plugins/algorithms/CompareSampleLogs.py]: Load the CNCS_7860 from doc tests. Run `FilterBadPulses` to eliminate al the  values when the accelerator was down. The `proton_charge` log is in picoCoulombs, with an average of 1.81E7, and a standard deviation of about 82000. It all depends on the units. I would just leave a lower bound

**Comment by AndreiSavici** [comment]
What would happen if one of the properties is NaN?
```
import numpy as np
ws1.mutableRun().addProperty('chopper_speed',np.nan,True)
```

**Comment by mganeva** [comment]
@AndreiSavici Good question. I will check it on Monday. An remove the upper bound too.

**Comment by mganeva** [comment]
OSX build has failed due to the "SliceViewerMantidPlotTest_SliceViewerPythonInterfaceTest" and the windows build has failed due to the "AlgorithmsTest_PDCalibrationTest". May I ask someone to restart these bulds?

**Pull request merged**